File: Data Structures & Algorithms/same-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        
        def preorder(node):
            res = []

            def dfs(node):
                if not node:
                    res.append(None)
                    return
                res.append(node.val)      # 👈 root FIRST
                dfs(node.left)
                dfs(node.right)

            dfs(node)
            return res

        return preorder(p) == preorder(q)

            

File: Data Structures & Algorithms/combinations/submission-5.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        vector<int> curset;
        DFS(1 ,n ,k , curset , res);
        return res;
        
    }
private:
    void DFS(int i ,int n , int k , vector<int>& curset , vector<vector<int>>& res){
        if (curset.size() == k){
            res.push_back(curset);
            return;
        }
        if (i > n){
            return;
        }
        curset.push_back(i);
        DFS(i +1 ,n,k , curset , res);
        curset.pop_back();

        DFS(i +1 ,n ,k , curset , res);
    }
};